src/fixup/fixup.cpp

```diff
@@ -645,6 +664,156 @@ TaskStatus ApplyFloorsImpl(T *rc, IndexDomain domain = IndexDomain::entire) {
+              pack(b, pblo, k, j, i) +=
+                  (-(A(b, j, i) - A(b, j + 1, i) + A(b, j, i + 1) - A(b, j + 1, i + 1)) /
+                   (2.0 * coords.CellWidthFA(X2DIR, k, j, i) * gammadet));
+              pack(b, pblo + 1, k, j, i) +=
+                  ((A(b, j, i) + A(b, j + 1, i) - A(b, j, i + 1) - A(b, j + 1, i + 1)) /
+                   (2.0 * coords.CellWidthFA(X1DIR, k, j, i) * gammadet));
+
+              // Used by ReduceMagneticFluxPhi
+              pack(b, cblo, k, j, i) = pack(b, pblo, k, j, i) * gammadet;
+              pack(b, cblo + 1, k, j, i) = pack(b, pblo + 1, k, j, i) * gammadet;
+            });
+
+        // Measure change in phi
+        const Real fiducial_phi_proc = History::ReduceMagneticFluxPhi(md);
+        const Real fiducial_phi = reduction::Sum(fiducial_phi_proc) / std::sqrt(mdot);
+        phi_factor = dphi / (fiducial_phi - phi) / dt;
+
+        // Remove the fiducial contribution
+        parthenon::par_for(
+            parthenon::LoopPatternMDRange(),
+            "Phoebus::Fixup::EndOfStepModify::EvaluateBField", DevExecSpace(), 0,
+            pack.GetDim(5) - 1, kb.s, kb.e, jb.s, jb.e - 1, ib.s, ib.e - 1,
+            KOKKOS_LAMBDA(const int b, const int k, const int j, const int i) {
+              const auto &coords = pack.GetCoords(b);
+              const Real gammadet = geom.DetGamma(CellLocation::Cent, b, k, j, i);
+
+              pack(b, pblo, k, j, i) -=
+                  (-(A(b, j, i) - A(b, j + 1, i) + A(b, j, i + 1) - A(b, j + 1, i + 1)) /
+                   (2.0 * coords.CellWidthFA(X2DIR, k, j, i) * gammadet));
+              pack(b, pblo + 1, k, j, i) -=
+                  ((A(b, j, i) + A(b, j + 1, i) - A(b, j, i + 1) - A(b, j + 1, i + 1)) /
+                   (2.0 * coords.CellWidthFA(X1DIR, k, j, i) * gammadet));
+
+              pack(b, cblo, k, j, i) = pack(b, pblo, k, j, i) * gammadet;
+              pack(b, cblo + 1, k, j, i) = pack(b, pblo + 1, k, j, i) * gammadet;
+            });
+      }
+
+      // Properly update B field
+      parthenon::par_for(
+          parthenon::LoopPatternMDRange(),
+          "Phoebus::Fixup::EndOfStepModify::EvaluateBField", DevExecSpace(), 0,
+          pack.GetDim(5) - 1, kb.s, kb.e, jb.s, jb.e, ib.s, ib.e,
+          KOKKOS_LAMBDA(const int b, const int k, const int j, const int i) {
+            const auto &coords = pack.GetCoords(b);
+            const Real gammadet = geom.DetGamma(CellLocation::Cent, b, k, j, i);
+
+            pack(b, pblo, k, j, i) +=
+                phi_factor * dt *
+                (-(A(b, j, i) - A(b, j + 1, i) + A(b, j, i + 1) - A(b, j + 1, i + 1)) /
+                 (2.0 * coords.CellWidthFA(X2DIR, k, j, i) * gammadet));
+            pack(b, pblo + 1, k, j, i) +=
+                phi_factor * dt *
+                ((A(b, j, i) + A(b, j + 1, i) - A(b, j, i + 1) - A(b, j + 1, i + 1)) /
+                 (2.0 * coords.CellWidthFA(X1DIR, k, j, i) * gammadet));
+          });
```

Comment:
Why can't we combine these two kernels? `A` doesn't change so feels like fiducial B-field and update can be done in the same loop?

Reply:
The former kernel is only being run when dphi/dt is being recalculated (typically only once every many timesteps). I think the former kernel can be bundled into the initialization step anyway and only done once.